### PR TITLE
feat(x/ecocredit)!: migrate add-credit-type proposal to msg based proposal

### PR DESCRIPTION
**Pull request by aleem1314**
## Description

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/regen/ecocredit/v1/tx.proto [hunk 1]
```diff
@@ -111,8 +112,23 @@ service Msg {
...
 }
 
+// MsgAddCreditType is the Msg/AddCreditType request type.
+message MsgAddCreditType {
+  // authority is the address of the governance account.
+  string authority = 1;
+
+  // credit_type defines a credit type to add to the credit types parameter. 
+  repeated CreditType credit_type = 2;
+}
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1/tx.proto:127]: I think we should only allow one credit type at a time. This will not be frequent and make things a bit simpler.

**Review by ryanchristo — COMMENTED** [review]
Looking good although I think we should only allow adding one credit type at a time. This would make the implementation a bit simpler. cc @clevinson thoughts?

**Review by ryanchristo — COMMENTED** [review]
Looking good. My bad for not generating proto after making formatting fixes in #951. Mostly small nits from me. Nice work taking on writing feature files!

File: x/ecocredit/server/core/features/msg_add_credit_type.feature [hunk 2]
```diff
@@ -0,0 +1,74 @@
+Feature: Msg/AddCreditType
```
Comment (technicallyty) [inline x/ecocredit/server/core/features/msg_add_credit_type.feature:1]: i think we also need the test case for attempting to add credit type with duplicate abbreviation. the abbreviation is the primary key, name is the unique index

File: x/ecocredit/server/core/msg_add_credit_type.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	}
+
+	found, err := k.stateStore.CreditTypeTable().HasByName(ctx, req.CreditType.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if found {
+		return nil, sdkerrors.ErrConflict.Wrapf("credit type with %s name already exists", req.CreditType.Name)
+	}
```
Comment (technicallyty) [inline x/ecocredit/server/core/msg_add_credit_type.go:26]: we should either do this for both name and abbreviation, or just attempt the `insert` call and check the returned error for which type of failure occurred so we can supply a more descriptive error message.

for duplicate name, you'd get an unique key violation error, and for duplicate abbreviation you'll get PrimaryKeyConstraint error. so we can either check for these or just do the pattern you've done above with the addition of abbreviation check

**Review by ryanchristo — APPROVED** [review]
[tACK](https://hackmd.io/EIWwUO_aRSa6Zh89LRRDbQ). Nice work on adding gherkin/gocuke tests.

**Review by technicallyty — APPROVED** [review]
looks good 👍🏻
